### PR TITLE
interfaces: add microstack-support interface

### PR DESCRIPTION
**Pull request by dshcherb**
Add an interface to enable MicroStack to work in a confined environment.

### CODE REVIEW

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:156]: This is allowed by the default template.

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:162]: Please plugs hardware-observe for all of these.

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:165]: This is already allowed by the default template via the 'consoles' abstraction.

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:175]: This is discussed in https://github.com/snapcore/snapd/pull/8909 where I mention the FHS says all the files for console locking should be `LCK..<something>` so that different processes can coordinate (see https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s09.html). Since putting in snap-specific directories is against the FHS and makes it difficult to actually coordinate, please adjust this to be:

```
/run/lock/ r,
/run/lock/LCK.._pts_* rwk,
```

(tested locally). If `/run/lock/LCK.._pts_*` is not sufficient, please state examples.

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:178]: This doesn't seem right. With microstack, there should be no process running with a libvirtd label. The default template via the base abstraction already has `unix peer=(label=@{profile_name}),` so I'm confused by this rule. Can you comment why it is needed?

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:179]: Under microstack, the libvirtd profile should be `snap.microstack.<something>` and the default template already has:
```
  # Allow apps from the same package to communicate with each other via an
  # abstract or anonymous socket
  unix (bind, listen) addr="@snap.@{SNAP_INSTANCE_NAME}.**",
  unix peer=(label=snap.@{SNAP_INSTANCE_NAME}.*),
```

so I think this can be removed?

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:191]: With these 3 capabilities, are you running the VMs as root? As 'snap_daemon'? (See https://forum.snapcraft.io/t/system-usernames/13386). It would be nice if microstack started them as non-root as 'snap_daemon:snap_daemon'. If you did that, you can remove these since they would be included in the template.

If not, what are you chowning and setuid/setgid'ing to?

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:194]: harrdware-observe and network-control both give sys_admin.

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:196]: Please plugs network-control.

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:197]: `man capabilities` states this is 'unused'. Can you comment? It probably belongs in network-control if it can be justified.

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:201]: It isn't surprising that nova isn't setting its directories and dropping/raising in a way to avoid these...

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:208]: Please plugs netlink-audit.

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:210]: This is in the default template.

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:213]: Please plugs process-control.

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:216]: Depending on what application is using this, you will get this by plugging network-control. If the app in your snap doesn't need it, we'll need to add it somewhere else (perhaps here). What is using it?

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:220]: I'm a bit surprised by this rule. Can you provide more details? Ie, I expected this rule:
```
change_profile -> libvirt-[0-9a-f]*-[0-9a-f]*-[0-9a-f]*-[0-9a-f]*-[0-9a-f]*,
```

Did you patch libvirt to use `microstack` instead of `libvirt` in its policy? If so, why wouldn't a rule like this work:
```
change_profile -> microstack-[0-9a-f]*-[0-9a-f]*-[0-9a-f]*-[0-9a-f]*-[0-9a-f]*,
```

?

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:222]: Similarly, are these for interacting with VM profiles?

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:225]: Please plugs 'network'.

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:228]: This is already allowed in network-control.

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:246]: You get all these already with system-usernames (see above).

File: interfaces/builtin/microstack_support.go
Comment (jdstrand) [inline interfaces/builtin/microstack_support.go:252]: tun is already done in network-control. I believe you need corresponding entries for vfio and nbd. I'm not sure what you want for both OTOH. I suggest creating the devices you want libvirt to have access to then using `udevadm info /dev/<thing>` and adjusting this accordingly.

**Review by jdstrand — DISMISSED** [review]
Thanks for submitting this. Lots of things to change, but much is just removing things that exist in other interfaces. Please do see various questions inline.

**Comment by dshcherb** [comment]
Rebased on top of recent snapd changes and added rules for the LVM support.

One more thing besides the LVM support that's needed for MicroStack to have an LVM-backed volume backend is running tgtd. This has to do with the way Cinder exposes access to LVM volumes to VMs - it is done via iSCSI. There are multiple reasons why that's the case:

* a VM might be live migrated from a host with the LVM volume but the volume will stay at the old node;
* LVM volumes may be created on a storage node so they need to be accessed over a network.

In the absence of a better volume backend, LVM provides a way to test volume creation which is what is needed to improve MicroStack gate tests since certain tempest tests require volume-backed instances to be created.

I am going to check which rules will be needed for running tgtd and if the interface will need to be modified further to make that work.

**Comment by dshcherb** [comment]
Added the iSCSI bits in https://github.com/snapcore/snapd/pull/8926/commits/4db6c1a02f5cedbfd878776424ec3ad74de7551b.

Unfortunately, upstream OpenStack does not currently allow one to leverage the QEMU's built-in iscsi support, therefore, MicroStack has to include `iscsid` in addition to `tgtd`.

Using `iscsid` is certainly more complicated as it uses the kernel data plane via the `iscsi_tcp` kernel module which means that snap removal becomes a more complex exercise involving the iSCSI session cleanup and that the `microstack_support` interfaces needs to be much more privileged in terms of `sysfs` access.

**Comment by dshcherb** [comment]
For reference, when tested with snapd built with this PR applied and a test-time workaround for [LP #1892895](https://bugs.launchpad.net/snapd/+bug/1892895) applied, I managed to get 0 failures of Refstack tests.

https://review.opendev.org/#/c/738242/9
https://zuul.opendev.org/t/openstack/build/19f8f05e192b471cbdb24c8d48467c47/log/job-output.txt#31891
```
2020-08-26 12:05:37.874765 \| ubuntu-bionic \| Ran: 104 tests in 629.866 sec.
2020-08-26 12:05:37.874779 \| ubuntu-bionic \|  - Success: 102
2020-08-26 12:05:37.874818 | ubuntu-bionic |  - Failures: 0
```
This is to show the applicability of the interface code to the target workload.

File: interfaces/builtin/microstack_support.go
Comment (mardy) [inline interfaces/builtin/microstack_support.go:65]: This is quite a dangerous rule. Can we somehow play with openstack or LVM so that the virtual group directories are created with names following a certain pattern?

File: interfaces/builtin/microstack_support.go
Comment (mardy) [inline interfaces/builtin/microstack_support.go:79]: You can write this as a single line: `/sys/kernel/iommu_groups/{,**} r,`

File: interfaces/builtin/microstack_support.go
Comment (mardy) [inline interfaces/builtin/microstack_support.go:162]: Don't you get the same effect if you modify the line above to use the `[a-i]` pattern? I see that in this way you are excluding `sdiz` -- is it some kind of special device?

File: interfaces/builtin/microstack_support.go
Comment (mardy) [inline interfaces/builtin/microstack_support.go:164]: Can you please add a comment describing why this is needed?

File: interfaces/builtin/microstack_support.go
Comment (mardy) [inline interfaces/builtin/microstack_support.go:165]: I guess you can use `@{PROC}` here too, and maybe move this up, where the other proc rules are.

**Review by mardy — COMMENTED** [review]
I've written some comments inline.

**Comment by mardy** [comment]
Hi again :-) I've been thinking a bit about this interface, and it seems to me that, no matter how carefully we write it, we are always leaving the door open for a process running inside the snap to break out of the confinement, if we allow libvirt to create apparmor profiles and run new processes with the newly created profiles. That's because the process running libvirt could be compromised, or there might be bugs that make it so that the generated apparmor profiles are way more permissive than what we'd like them to be.

This would not be an issue if we could rely on the `NoNewPrivileges` flag, because then we would be sure that only profile transitions reducing the permission set would be allowed; unfortunately, though, for reasons that I've yet to fully grasp, we are not setting that flag in `snap-confine`.

So, I wonder if it wouldn't be better, until we have the `NoNewPrivileges` flag set, to disable apparmor in libvirt and have all processes inside this snap running under the same AppArmor profile. It's a suboptimal solution, but at least it would guarantee us that no process from inside this snap can escape the confinement (as long as we don't have cap_sys_admin enabled).

**Comment by jdstrand** [comment]
> Hi again :-) I've been thinking a bit about this interface, and it seems to me that, no matter how carefully we write it, we are always leaving the door open for a process running inside the snap to break out of the confinement, if we allow libvirt to create apparmor profiles and run new processes with the newly created profiles. That's because the process running libvirt could be compromised, or there might be bugs that make it so that the generated apparmor profiles are way more permissive than what we'd like them to be.
> 
> This would not be an issue if we could rely on the `NoNewPrivileges` flag, because then we would be sure that only profile transitions reducing the permission set would be allowed; unfortunately, though, for reasons that I've yet to fully grasp, we are not setting that flag in `snap-confine`.
> 
> So, I wonder if it wouldn't be better, until we have the `NoNewPrivileges` flag set, to disable apparmor in libvirt and have all processes inside this snap running under the same AppArmor profile. It's a suboptimal solution, but at least it would guarantee us that no process from inside this snap can escape the confinement (as long as we don't have cap_sys_admin enabled).

NoNewPrivileges interferes with profile transitions by snap-confine which is why we do not set it there. There has been some work in this area upstream but it is not yet available everywhere. As for setting it after snap-confine has transitioned, because of the interactions between historic apparmor and NNP, this can alter the behavior of the applications that try to use NNP themselves, etc. We've felt that the combination of the existing sandbox technologies (apparmor, seccomp, etc) achieved many of the same goals as NNP but with added flexibility for (privileged) interfaces that need it.

As for the observation about libvirt, you are correct. If an application can load apparmor policy and transition into it, it can trivially break out of confinement (note, libvirt (and docker, containerd, lxd, etc, etc) have many more ways to break out (consider mounting a raw disk that would allow writes that changed policy, apparmor_parser/snapd itself, etc), so how we've decided to handle that is to called these 'superprivileged' interfaces, where the interfaces allow the application to do their job but we disallow installation/connection/etc through snapd interface policy in the snap declarations. In this manner, we might grant microk8s-support to Canonical since it is the upstream for microk8s, but disallow it to everyone else. We then document that the interface is advisory and grants ownership to the system.

In general, IMHO approaching interfaces that require lots of privilege can be done in phases. The first phase is like I describe above (lenient policy that let's them get the job done, with some guardrails (eg, let them only transition to policy with a certain naming convention, etc)). As technologies mature, later phases could further restrict (where the interface could have attributes that allow choosing the 'flavor' of the interface policy, and these attributes are mediatable in the snap declaration) and take advantage of improvements to sandboxing technologies. It's also possible to make adjustments to the internal components of the snap to understand snap sandboxing (eg, in the case of libvirt, perhaps there is a snap helper that libvirtd could call out to do policy writes/transitions or libvirt is modified to have reduced functionality when running in a snap sandbox (ie, it can't or can only attach a subset of devices, etc)).

CC @alexmurray for awareness

File: interfaces/builtin/microstack_support.go
Comment (pedronis) [inline interfaces/builtin/microstack_support.go:232]: cc @anonymouse64 

would some/all of the services here need Delegate=yes ?

**Comment by pedronis** [comment]
> ... could be extracted into separate -control or -support interfaces

there is something going in that direction here #9585

File: interfaces/builtin/microstack_support.go
Comment (dshcherb) [reply interfaces/builtin/microstack_support.go:65]: Volume groups are created using `vgcreate` which uses has a static template for naming a directory that would contain LV symlinks for a particular VG: `/dev/<vgname>`

https://github.com/lvmteam/lvm2/blob/v2_03_07/lib/activate/fs.c#L42-L45
https://github.com/lvmteam/lvm2/blob/v2_03_07/lib/activate/fs.c#L275-L279

LVM volume group names are configurable in OpenStack and users may configure as many instances of this volume driver as they want (we have an experimental loop-based backend for testing that has a specific name but there are no restrictions as to how users can configure their own one).

https://docs.openstack.org/cinder/latest/configuration/block-storage/drivers/lvm-volume-driver.html#id1
volume_group = <vg-name>

The usage scenario of the LVM driver for MicroStack is to provide an ability for a user to configure storage as they want it and provide a VG on top of it for use by MicroStack. So restricting VG names would be limiting the way users might want to use or name things.

To summarize: the issue is mostly that LVM has a static location that it wants to be able to access and there are no means to control it.

**Review by dshcherb — COMMENTED** [review]
Sorry about the long delay in response. Answered some in-line comments in the review.

File: interfaces/builtin/microstack_support.go
Comment (anonymouse64) [reply interfaces/builtin/microstack_support.go:232]: @dshcherb yes if you want to add that, I think that is appropriate here, take a look at the docker-support interface for an example of how to do that

File: interfaces/builtin/microstack_support.go
Comment (mardy) [reply interfaces/builtin/microstack_support.go:65]: Plan A would be to modify the LVM tools to use a different pattern for the names (like, change that format string to be `%s/vg_%s`); I guess the tools are part of the snap, so we could patch them.

Or, in alternative:

> The usage scenario of the LVM driver for MicroStack is to provide an ability for a user to configure storage as they want it and provide a VG on top of it for use by MicroStack. So restricting VG names would be limiting the way users might want to use or name things.

Yes, but isn't this a compromise we could make? I mean, in the apparmor profile we could allow those names which are used by the most popular MicroStack services, and also allow a certain pattern for user chosen names (for instance, that volume groups start with `vg`). As long as we find a way to document this, do you still see an issue?

File: interfaces/builtin/microstack_support.go
Comment (mardy) [reply interfaces/builtin/microstack_support.go:163]: OK, seems legit :-)

File: interfaces/builtin/microstack_support.go
Comment (dshcherb) [reply interfaces/builtin/microstack_support.go:65]: > do you still see an issue

For this interface, yes, since the MicroStack snap also gets `block_devices` interface access which means `rw` [access](https://github.com/snapcore/snapd/blob/2.51.2/interfaces/builtin/block_devices.go#L60-L65) to all block devices. As such, it is free to overwrite any signatures or contents on the backing block devices, including the data of any volumes in any of the volume groups. This makes the volume group name restriction artificial since MicroStack is the only user of the `microstack-support` interface.

>  I guess the tools are part of the snap, so we could patch them.

I went through the pain of upgrading the snap to core20 and newer versions of different parts (as a part of changing the base series and OpenStack version) and a significant amount of time was spend on rebasing/reworking all the patches we carry already - it would be preferable to avoid even more patches since we do not have capacity to carry this forward all the time.

It would be better to focus on allowing libvirt running as a part of MicroStack to apply AppArmor policies to QEMU processes it spawns but this does not work at the moment due to a different issue https://forum.snapcraft.io/t/call-chains-loose-ld-library-path-and-thereby-child-programs-fail/24382. We never had this enabled in MicroStack so it can be an incremental step.

> allow those names which are used by the most popular MicroStack services

That's the thing - OpenStack does not configure the LVM backend itself. It is completely up to the deployer which VG name gets chosen and there may be preexisting VGs that people might want to use.

In summary: because MicroStack has enough privileges to circumvent the vg name restriction you are talking about, I would not create artificial restrictions which could later result in bug reports. LXD does not have such restrictions while it [can work with LVM](https://github.com/lxc/lxd/blob/lxd-4.16/lxd/storage/drivers/driver_lvm.go) and in this case we are dealing with the hypervisor software so it would make sense for it to be able to use arbitrary VG names.

**Comment by dshcherb** [comment]
I just did another functional test run (Tempest, [refstack subset](https://github.com/openstack-snaps/microstack-test/blob/master/2020.06-test-list.txt)) with locally built snapd that had `microstack-support` included.

All volume tests failed due to https://bugs.launchpad.net/snapd/+bug/1892895 which is not something I can fix in this interface (see the details in the bug, including [comment #4](https://bugs.launchpad.net/snapd/+bug/1892895/comments/4)).

When I manually allowed the necessary devices via wildcard rules to avoid the race condition like this

```
sudo bash -c 'echo b 253:* rwm > /sys/fs/cgroup/devices/snap.microstack.<service-name>/devices.allow'
```

I was able to make to get a full pass: https://paste.ubuntu.com/p/TB6VtxThSz/

In summary: I do not see any new rules that need to be added to this interface in order for us to get our functional tests to pass (except for volume tests for experimental/optional volume functionality).

File: interfaces/builtin/microstack_support.go
Comment (mardy) [inline interfaces/builtin/microstack_support.go:240]: minor: I would break this line into multiple lines (one per module), and maybe add a comment next to those modules whose name is not very explicative, like `target-core-mod` (which seems related to SCSI).

File: interfaces/builtin/microstack_support_test.go
Comment (mardy) [inline interfaces/builtin/microstack_support_test.go:91]: Please add a couple of more lines to this check, preferably those which are microstack-specific (`/dev/microstack-*/{,**} rw,` sounds like a good candidate)

**Review by mardy — COMMENTED** [review]
Looks fine to me, just minor comments. Given that the interface is relatively trivial in logic (it's complex as far as AppArmor rules are concerned, but it's a very typical interface as far as snapd is concerned) I wouldn't insist on a spread test, but let's see what other people think.

File: interfaces/builtin/microstack_support_test.go
Comment (dshcherb) [reply interfaces/builtin/microstack_support_test.go:91]: Done.

File: interfaces/builtin/microstack_support.go
Comment (dshcherb) [reply interfaces/builtin/microstack_support.go]: Done.

File: interfaces/builtin/microstack_support_test.go
Comment (dshcherb) [reply interfaces/builtin/microstack_support_test.go:91]: Done https://github.com/snapcore/snapd/pull/8926/commits/d84dd2f166a3eac250c1e44d14f1b5d31e25f363

**Review by mardy — APPROVED** [review]
Excellent! Let's wait, though, for the security team review (as well for some more expert team-mates of mine). :-)

File: interfaces/builtin/microstack_support.go
Comment (alexmurray) [inline interfaces/builtin/microstack_support.go:114]: ` @{PROC}/cmdline r,` is in the base template - this can be removed.

File: interfaces/builtin/microstack_support.go
Comment (alexmurray) [inline interfaces/builtin/microstack_support.go:127]: `/dev/ptmx rw` is in the base template - this can be removed.

File: interfaces/builtin/network_control.go
Comment (alexmurray) [inline interfaces/builtin/network_control.go:192]: This is already covered by the `firewall-control` interface - I don't think it is appropriate for `network-control` to gain this as it seems too much like scope creep compared to what it currently covers.

**Review by alexmurray — APPROVED** [review]
Overall this looks pretty solid - the only glaring bit which stood out was the iptables related change to the network-control interface - this doesn't seem right to me, but otherwise this seems fine as is (whilst I like the idea of a libvirt-support or similar interface which could encapsulate some of this, I don't think it is fair to block this PR on waiting for that, especially given we don't seem to have other users for such an interface at the moment).

**Comment by dshcherb** [comment]
I did another functional test run after https://github.com/snapcore/snapd/pull/8926/commits/ebc2f4f1c1d4f87c160a382951816985dce0af95 (without volume testing since those rules were not changed). Results: https://paste.ubuntu.com/p/rPtZFv6DVg/.

Unit tests are passing and most of the spread tests too (the failures are not related to this change).

File: interfaces/builtin/microstack_support_test.go
Comment (anonymouse64) [inline interfaces/builtin/microstack_support_test.go:57]: nitpick, but nowadays we have `MockConnectedSlot` and `MockConnectedPlug` for getting the Slot and SlotInfo, etc. in an easier way

File: interfaces/builtin/microstack_support.go
Comment (anonymouse64) [inline interfaces/builtin/microstack_support.go:149]: this should be `/{var/,}run/...` shouldn't it? since /var/run is a symlink to /run?

**Review by anonymouse64 — APPROVED** [review]
lgtm, one question about non-systemd layouts and a test suggestion

File: interfaces/builtin/microstack_support.go
Comment (dshcherb) [reply interfaces/builtin/microstack_support.go:105]: This was discussed above earlier (see https://github.com/snapcore/snapd/pull/8926#discussion_r449584383 for an in-depth explanation with code references) - it has to do with presence of `systemd-machined` that comes with the `systemd-container` on Ubuntu (not present in our cloud images by default).

Since `systemd-machined` isn't present, the generic layout is used on a system with systemd present.

Hence those rules are here.

File: interfaces/builtin/microstack_support.go
Comment (dshcherb) [reply interfaces/builtin/microstack_support.go:149]: Ack, ty,

File: interfaces/builtin/microstack_support.go
Comment (anonymouse64) [reply interfaces/builtin/microstack_support.go:105]: Ah thanks for the pointer, I think I had read that comment originally a few weeks/months ago but forgot about it when I did another review again

File: interfaces/builtin/microstack_support_test.go
Comment (dshcherb) [reply interfaces/builtin/microstack_support_test.go]: Ack, changed to those in https://github.com/snapcore/snapd/pull/8926/commits/bea48764d3579a0c8a7d4f2d3ffa8f0819d8a8f3

Comment (anonymouse64) [reply interfaces/builtin/microstack_support_test.go]: thanks

**Comment by dshcherb** [comment]
Hey @jdstrand, any chance you could approve this or dismiss the change request? I think what was requested originally has been addressed but the change request is still active.

Thanks in advance!

**Pull request merged**
